pkg/appgw/configbuilder_test.go

```diff
@@ -0,0 +1,396 @@
+	}
+
+	// Create the mock K8s client.
+	k8sClient := testclient.NewSimpleClientset()
+	_, err := k8sClient.CoreV1().Namespaces().Create(ns)
+	It("should have not failed", func() { Expect(err).ToNot(HaveOccurred()) })
+
+	_, err = k8sClient.CoreV1().Nodes().Create(node)
+	It("should have not failed", func() { Expect(err).ToNot(HaveOccurred()) })
+
+	_, err = k8sClient.ExtensionsV1beta1().Ingresses(ingressNS).Create(ingress)
+	It("should have not failed", func() { Expect(err).ToNot(HaveOccurred()) })
+
+	_, err = k8sClient.CoreV1().Services(ingressNS).Create(service)
+	It("should have not failed", func() { Expect(err).ToNot(HaveOccurred()) })
+
+	_, err = k8sClient.CoreV1().Endpoints(ingressNS).Create(endpoints)
+	It("should have not failed", func() { Expect(err).ToNot(HaveOccurred()) })
+
+	_, err = k8sClient.CoreV1().Pods(ingressNS).Create(pod)
+	It("should have not failed", func() { Expect(err).ToNot(HaveOccurred()) })
+
+	crdClient := fake.NewSimpleClientset()
+	istioCrdClient := istio_fake.NewSimpleClientset()
+	ctxt := k8scontext.NewContext(k8sClient, crdClient, istioCrdClient, []string{ingressNS}, 1000*time.Second)
+
+	appGwy := &n.ApplicationGateway{}
+	// Since this is a mock the `Application Gateway v2` does not have a public IP. During configuration process
+	// the controller would expect the `Application Gateway v2` to have some public IP before it starts generating
+	// configuration for the application gateway, hence creating this dummy configuration in the application gateway configuration.
+	appGwy.ApplicationGatewayPropertiesFormat = &n.ApplicationGatewayPropertiesFormat{
+		FrontendIPConfigurations: &[]n.ApplicationGatewayFrontendIPConfiguration{
+			{
+				Name: to.StringPtr("*"),
+				Etag: to.StringPtr("*"),
+				ID:   to.StringPtr("*"),
+				ApplicationGatewayFrontendIPConfigurationPropertiesFormat: &n.ApplicationGatewayFrontendIPConfigurationPropertiesFormat{
+					PublicIPAddress: &n.SubResource{
+						ID: to.StringPtr("x/y/z"),
+					},
+				},
+			},
+		},
+	}
+
+	// Initialize the `ConfigBuilder`
+	configBuilder := NewConfigBuilder(ctxt, &appGwIdentifier, appGwy, record.NewFakeRecorder(100))
+
+	Context("Tests Application Gateway config creation", func() {
+		cbCtx := &ConfigBuilderContext{
+			IngressList:  []*v1beta1.Ingress{ingress},
+			ServiceList:  serviceList,
+			EnvVariables: environment.GetFakeEnv(),
+		}
+
+		It("Should have created correct App Gateway config JSON blob", func() {
+			appGW, err := configBuilder.Build(cbCtx)
+			Expect(err).ToNot(HaveOccurred())
+
+			jsonBlob, err := appGW.MarshalJSON()
+			Expect(err).ToNot(HaveOccurred())
+
+			var into map[string]interface{}
+			err = json.Unmarshal(jsonBlob, &into)
+			Expect(err).ToNot(HaveOccurred())
+
+			jsonBlob, err = json.MarshalIndent(into, "--", "    ")
+			Expect(err).ToNot(HaveOccurred())
+
+			jsonTxt := string(jsonBlob)
+
+			expected := `{
+--    "properties": {
+--        "backendAddressPools": [
+--            {
```

Comment:
I think this backend pool is missing a backend pool. Maybe something is wrong in the setup ?